Exemplo de Auto Machine Learning:

Vamos construir um modelo de AML e um de Random Forest para compara-los

In [2]:
import h2o
from h2o.automl import H2OAutoML
from h2o.estimators import H2ORandomForestEstimator
import pandas as pd

In [3]:
h2o.init()

Checking whether there is an H2O instance running at http://localhost:54321 . connected.


--------------------------  ---------------------------------------------------------
H2O_cluster_uptime:         23 mins 55 secs
H2O_cluster_timezone:       America/Sao_Paulo
H2O_data_parsing_timezone:  UTC
H2O_cluster_version:        3.30.1.2
H2O_cluster_version_age:    15 days
H2O_cluster_name:           H2O_started_from_R_Giovanni_hsp778
H2O_cluster_total_nodes:    1
H2O_cluster_free_memory:    1.871 Gb
H2O_cluster_total_cores:    4
H2O_cluster_allowed_cores:  4
H2O_cluster_status:         locked, healthy
H2O_connection_url:         http://localhost:54321
H2O_connection_proxy:       {"http": null, "https": null}
H2O_internal_security:      False
H2O_API_Extensions:         Amazon S3, Algos, AutoML, Core V3, TargetEncoder, Core V4
Python_version:             3.7.4 final
--------------------------  ---------------------------------------------------------

In [4]:
#Importa dados 
imp = pd.read_csv('Churn_treino.csv', sep=";")
imp.head()

CreditScore Geography  Gender  Age  Tenure   Balance  NumOfProducts  \
0          619    France  Female   42       2         0              1   
1          608     Spain  Female   41       1   8380786              1   
2          502    France  Female   42       8   1596608              3   
3          699    France  Female   39       1         0              2   
4          850     Spain  Female   43       2  12551082              1   

   HasCrCard  IsActiveMember  EstimatedSalary  Exited  
0          1               1         10134888       1  
1          0               1         11254258       0  
2          1               0         11393157       1  
3          0               0          9382663       0  
4          1               1           790841       0

In [5]:
# divide em treino e teste
imp = h2o.H2OFrame(imp)
treino,teste = imp.split_frame(ratios=[.7])
    # ratios = .7 = 70% dos dados são de treino

Parse progress: |█████████████████████████████████████████████████████████| 100%


In [6]:
#Transforma a variável dependente em fator
treino["Exited"] = treino["Exited"].asfactor()
teste["Exited"] = teste["Exited"].asfactor()

Modelo de AML:

In [7]:
modelo_automl = H2OAutoML(max_runtime_secs=180, sort_metric = 'AUC')
    # max_runtime_secs = tempo máximo de busca em uma métrica, 180s
    # sort_metric = métrica de avaliação
modelo_automl.train( y="Exited", training_frame=treino)
    # Treinamento, y = classe

AutoML progress: |█
00:52:25.312: AutoML: XGBoost is not available; skipping it.

███████████████████████████████████████████████████████| 100%


Modelo Random Forest:

In [8]:
model_floresta = H2ORandomForestEstimator(nfolds=5)
model_floresta.train( y="Exited", training_frame=treino)

drf Model Build progress: |███████████████████████████████████████████████| 100%


Ranqueando os melhores AML:

In [9]:
ranking = modelo_automl.leaderboard
ranking = ranking.as_data_frame()
ranking

model_id       auc   logloss  \
0    StackedEnsemble_AllModels_AutoML_20200919_005225  0.858196  0.348946   
1         GBM_grid__1_AutoML_20200919_005225_model_10  0.858052  0.345741   
2                        GBM_2_AutoML_20200919_005225  0.856561  0.345008   
3         GBM_grid__1_AutoML_20200919_005225_model_11  0.856434  0.346505   
4   StackedEnsemble_BestOfFamily_AutoML_20200919_0...  0.856279  0.351271   
5         GBM_grid__1_AutoML_20200919_005225_model_12  0.854963  0.347347   
6                        GBM_5_AutoML_20200919_005225  0.854642  0.347896   
7                        GBM_1_AutoML_20200919_005225  0.854515  0.345845   
8          GBM_grid__1_AutoML_20200919_005225_model_4  0.851986  0.352457   
9          GBM_grid__1_AutoML_20200919_005225_model_8  0.849735  0.353569   
10                       GBM_3_AutoML_20200919_005225  0.849510  0.352959   
11                       GBM_4_AutoML_20200919_005225  0.847464  0.356405   
12        GBM_grid__1_AutoML_20200919_005225_model_14  0.843142  0.357997   
13         GBM_grid__1_AutoML_20200919_005225_model_7  0.842768  0.362316   
14        GBM_grid__1_AutoML_20200919_005225_model_13  0.841701  0.361255   
15         GBM_grid__1_AutoML_20200919_005225_model_5  0.841066  0.368780   
16         GBM_grid__1_AutoML_20200919_005225_model_3  0.840429  0.363363   
17              DeepLearning_1_AutoML_20200919_005225  0.839850  0.366018   
18         GBM_grid__1_AutoML_20200919_005225_model_1  0.837880  0.377951   
19         GBM_grid__1_AutoML_20200919_005225_model_2  0.836104  0.367700   
20  DeepLearning_grid__1_AutoML_20200919_005225_mo...  0.835626  0.376274   
21                       DRF_1_AutoML_20200919_005225  0.835474  0.499889   
22                       XRT_1_AutoML_20200919_005225  0.833014  0.402783   
23  DeepLearning_grid__1_AutoML_20200919_005225_mo...  0.830328  0.381449   
24         GBM_grid__1_AutoML_20200919_005225_model_6  0.829352  0.378076   
25  DeepLearning_grid__2_AutoML_20200919_005225_mo...  0.817200  0.456672   
26                       GLM_1_AutoML_20200919_005225  0.759381  0.437176   
27        GBM_grid__1_AutoML_20200919_005225_model_15  0.723731  0.485178   
28         GBM_grid__1_AutoML_20200919_005225_model_9  0.719400  0.681516   

       aucpr  mean_per_class_error      rmse       mse  
0   0.697854              0.236597  0.324627  0.105383  
1   0.692713              0.239690  0.324279  0.105157  
2   0.695031              0.234922  0.323902  0.104912  
3   0.694639              0.239051  0.324547  0.105331  
4   0.690553              0.232004  0.325915  0.106221  
5   0.692014              0.230516  0.324925  0.105576  
6   0.689220              0.234714  0.325563  0.105991  
7   0.694180              0.235481  0.323843  0.104874  
8   0.682133              0.246994  0.327764  0.107429  
9   0.682387              0.236662  0.328123  0.107665  
10  0.682421              0.254508  0.328207  0.107720  
11  0.681236              0.236654  0.329604  0.108639  
12  0.677746              0.257883  0.329678  0.108688  
13  0.674645              0.253948  0.331089  0.109620  
14  0.667828              0.248560  0.332107  0.110295  
15  0.669615              0.252274  0.334397  0.111822  
16  0.664880              0.245266  0.332833  0.110778  
17  0.656830              0.249303  0.334603  0.111959  
18  0.665335              0.260262  0.335668  0.112673  
19  0.654793              0.236990  0.335734  0.112717  
20  0.651291              0.254812  0.337043  0.113598  
21  0.659148              0.250691  0.333457  0.111194  
22  0.653644              0.258315  0.336915  0.113512  
23  0.643778              0.255943  0.339818  0.115476  
24  0.643081              0.250423  0.339052  0.114957  
25  0.624267              0.258250  0.351928  0.123853  
26  0.457901              0.305605  0.372641  0.138861  
27  0.444326              0.325616  0.394088  0.155305  
28  0.451451              0.309988  0.417487  0.174295

Foram criados 28 modelos
O melhor foi: StackedEnsemble_AllModels_AutoML_20200919_005225
Que teve o AUC de 0.858

Performance do Random Forest:

In [10]:
performance = model_floresta.model_performance(test_data=teste)
performance


ModelMetricsBinomial: drf
** Reported on test data. **

MSE: 0.10206028895736116
RMSE: 0.3194687605343614
LogLoss: 0.3785837843553372
Mean Per-Class Error: 0.21811967727476467
AUC: 0.8570305945361975
AUCPR: 0.6854405048900383
Gini: 0.7140611890723949

Confusion Matrix (Act/Pred) for max f1 @ threshold = 0.32999999999999996: 


0      1   Error             Rate
0      0  2092.0  300.0  0.1254   (300.0/2392.0)
1      1   189.0  393.0  0.3247    (189.0/582.0)
2  Total  2281.0  693.0  0.1644   (489.0/2974.0)


Maximum Metrics: Maximum metrics at their respective thresholds


metric  threshold        value    idx
0                        max f1    0.33000     0.616471  151.0
1                        max f2    0.19000     0.687609  235.0
2                  max f0point5    0.56475     0.683003   70.0
3                  max accuracy    0.56475     0.868191   70.0
4                 max precision    1.00000     1.000000    0.0
5                    max recall    0.00000     1.000000  399.0
6               max specificity    1.00000     1.000000    0.0
7              max absolute_mcc    0.52000     0.535441   80.0
8    max min_per_class_accuracy    0.24000     0.778351  208.0
9   max mean_per_class_accuracy    0.29000     0.781880  176.0
10                      max tns    1.00000  2392.000000    0.0
11                      max fns    1.00000   577.000000    0.0
12                      max fps    0.00000  2392.000000  399.0
13                      max tps    0.00000   582.000000  399.0
14                      max tnr    1.00000     1.000000    0.0
15                      max fnr    1.00000     0.991409    0.0
16                      max fpr    0.00000     1.000000  399.0
17                      max tpr    0.00000     1.000000  399.0


Gains/Lift Table: Avg response rate: 19,57 %, avg score: 21,74 %


group  cumulative_data_fraction  lower_threshold      lift  \
0       1                  0.011432           0.9600  4.809379   
1       2                  0.025219           0.9200  4.860699   
2       3                  0.031607           0.9000  5.109966   
3       4                  0.040013           0.8408  4.905567   
4       5                  0.051446           0.8000  4.208207   
5       6                  0.100202           0.6135  3.418391   
6       7                  0.152320           0.4800  2.472564   
7       8                  0.208473           0.3800  1.499331   
8       9                  0.306658           0.2600  1.382491   
9      10                  0.402152           0.1800  0.719713   
10     11                  0.517485           0.1200  0.506527   
11     12                  0.627774           0.0800  0.436217   
12     13                  0.703093           0.0600  0.205311   
13     14                  0.879623           0.0200  0.214132   
14     15                  1.000000           0.0000  0.099916   

    cumulative_lift  response_rate     score  cumulative_response_rate  \
0          4.809379       0.941176  0.970588                  0.941176   
1          4.837434       0.951220  0.928740                  0.946667   
2          4.892520       1.000000  0.900877                  0.957447   
3          4.895261       0.960000  0.868350                  0.957983   
4          4.742582       0.823529  0.820902                  0.928105   
5          4.098261       0.668966  0.694436                  0.802013   
6          3.542007       0.483871  0.532901                  0.693157   
7          2.991802       0.293413  0.418133                  0.585484   
8          2.476540       0.270548  0.305555                  0.484649   
9          2.059367       0.140845  0.210100                  0.403010   
10         1.713283       0.099125  0.138898                  0.335283   
11         1.488924       0.085366  0.089692                  0.291377   
12         1.351416       0.040179  0.061517                  0.264467   
13         1.123177       0.041905  0.029680                  0.219801   
14         1.000000       0.019553  0.000641                  0.195696   

    cumulative_score  capture_rate  cumulative_capture_rate        gain  \
0           0.970588      0.054983                 0.054983  380.937942   
1           0.947711      0.067010                 0.121993  386.069902   
2           0.938245      0.032646                 0.154639  410.996564   
3           0.923561      0.041237                 0.195876  390.556701   
4           0.900748      0.048110                 0.243986  320.820699   
5           0.800361      0.166667                 0.410653  241.839080   
6           0.708846      0.128866                 0.539519  147.256402   
7           0.630541      0.084192                 0.623711   49.933123   
8           0.526488      0.135739                 0.759450   38.249070   
9           0.451359      0.068729                 0.828179  -28.028653   
10          0.381721      0.058419                 0.886598  -49.347279   
11          0.330416      0.048110                 0.934708  -56.378342   
12          0.301610      0.015464                 0.950172  -79.468888   
13          0.247037      0.037801                 0.987973  -78.586811   
14          0.217377      0.012027                 1.000000  -90.008447   

    cumulative_gain  kolmogorov_smirnov  
0        380.937942            0.054147  
1        383.743414            0.120321  
2        389.252029            0.152967  
3        389.526120            0.193786  
4        374.258249            0.239388  
5        309.826103            0.385987  
6        254.200709            0.481409  
7        199.180246            0.516270  
8        147.654036            0.562962  
9        105.936742            0.529684  
10        71.328283            0.458922  
11        48.892411            0.381614  
12        35.14

O Random Forest teve o AUC de AUC: 0.8570305945361975
O AML foi 0.858, levemente melhor 

Prevendo com AML:

In [11]:
teste = pd.read_csv('Churn_prever.csv', sep=";")
teste = h2o.H2OFrame(teste)
prever = modelo_automl.leader.predict(teste) # leader = usa o melhor modelo para previsão
prever

Parse progress: |█████████████████████████████████████████████████████████| 100%
stackedensemble prediction progress: |████████████████████████████████████| 100%


predict p0 p1 
 
 
 0 0.920235 0.0797651 
 0 0.926045 0.0739549 
 1 0.0394558 0.960544 
 0 0.943532 0.0564679 
 0 0.931241 0.0687586 
 0 0.906473 0.0935271 
 0 0.848754 0.151246 
 0 0.873545 0.126455 
 0 0.91723 0.0827702